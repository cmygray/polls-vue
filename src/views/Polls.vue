<template>
  <v-container tag="section" class="polls">
    <h1>All Polls</h1>

    <v-text-field v-model="titleToAppend" label="Poll title" />

    <v-btn
      large
      block
      color="primary"
      :disabled="!titleToAppend"
      @click="addPoll"
      >Create</v-btn
    >

    <poll-item
      v-for="poll in polls"
      :key="poll.id"
      :poll="poll"
      @poll-item:vote="handleVote"
    />
  </v-container>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { Action, Getter } from 'vuex-class';

import PollItem from '@/components/PollItem.vue';
import { Choice, Poll } from '@/models';

@Component({
  components: {
    PollItem,
  },
})
export default class Polls extends Vue {
  titleToAppend: string = '';

  @Getter('allPolls')
  polls: Poll[];

  @Action('fetchPolls')
  fetchPolls: () => Promise<Poll[]>;

  @Action('createPoll')
  createPoll: (param: string) => Promise<Poll>;

  @Action('voteToPoll')
  vote: (choice: Choice) => Promise<Poll>;

  async mounted() {
    await this.fetchPolls();
  }

  async addPoll() {
    if (!this.titleToAppend) return;

    try {
      const { id } = await this.createPoll(this.titleToAppend);

      this.$router.push({
        name: 'poll.edit',
        params: {
          id,
        },
      });
    } catch (error) {
      //
    }
  }

  async handleVote(choice: Choice) {
    if (!choice) return;

    try {
      await this.vote(choice);
    } catch (error) {
      //
    }
  }
}
</script>

<style lang="scss"></style>
