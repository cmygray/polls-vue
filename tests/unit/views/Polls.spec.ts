import Vuex from 'vuex';
import { createLocalVue, mount } from '@vue/test-utils';

import Polls from '@/views/Polls.vue';
import { pollList } from '../__fixtures__';
import '../setup';

const localVue = createLocalVue();
localVue.use(Vuex)

describe('Polls.vue', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = mount(Polls, {
      localVue,
      store: new Vuex.Store({
        actions: {
          fetchPolls: () => Promise.resolve()
        },
        getters: {
          allPolls: () => pollList
        }
      }),
    });
  });

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('on component mounted', () => {
    beforeEach(() => {
      wrapper = mount(Polls, {
        localVue,
        computed: {
          polls: () => pollList
        },
        mocks: {
          $store: {
            dispatch: jest.fn()
          }
        }
      });
    })
    it('should dispatch expected action', () => {
      expect(wrapper.vm.$store.dispatch).toBeCalledWith('fetchPolls')
    })
  })
});
