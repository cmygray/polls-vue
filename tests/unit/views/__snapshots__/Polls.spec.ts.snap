// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polls.vue renders as expected 1`] = `
<section class="container polls">
  <h1>All Polls</h1>
  <div class="v-input v-text-field theme--light">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">Poll title</label><input aria-label="Poll title" type="text"></div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper"></div>
        </div>
      </div>
    </div>
  </div> <button disabled="disabled" type="button" class="v-btn v-btn--block v-btn--disabled v-btn--large theme--light primary--text">
    <div class="v-btn__content">Create</div>
  </button>
  <article class="poll my-4 v-card v-sheet theme--light">
    <div class="v-card__title title">
      safdasfsdfaghf
    </div>
    <div class="v-card__text">
      <dl>
        <dt><small>시작</small></dt>
        <dd>2019-06-18 10:58</dd>
        <dt><small>종료</small></dt>
        <dd>
          2019-06-21 00:00
        </dd>
      </dl>
    </div>
  </article>
  <article class="poll my-4 v-card v-sheet theme--light">
    <div class="v-card__title title">
      TEST
    </div>
    <div class="v-card__text">
      <dl>
        <dt><small>시작</small></dt>
        <dd>2019-06-18 13:33</dd>
        <dt><small>종료</small></dt>
        <dd>
          2019-06-21 00:00
        </dd>
      </dl>
    </div>
  </article>
  <article class="poll my-4 v-card v-sheet theme--light">
    <div class="v-card__title title">
      TEST
    </div>
    <div class="v-card__text">
      <dl>
        <dt><small>시작</small></dt>
        <dd>2019-06-18 15:28</dd>
        <dt><small>종료</small></dt>
        <dd>
          2019-06-21 00:00
        </dd>
      </dl>
    </div>
  </article>
  <article class="poll my-4 v-card v-sheet theme--light">
    <div class="v-card__title title">
      TEST
    </div>
    <div class="v-card__text">
      <dl>
        <dt><small>시작</small></dt>
        <dd>2019-06-18 15:30</dd>
        <dt><small>종료</small></dt>
        <dd>
          2019-06-21 00:00
        </dd>
      </dl>
    </div>
  </article>
  <article class="poll my-4 v-card v-sheet theme--light">
    <div class="v-card__title title">
      TEST
    </div>
    <div class="v-card__text">
      <dl>
        <dt><small>시작</small></dt>
        <dd>2019-06-18 12:00</dd>
        <dt><small>종료</small></dt>
        <dd>
          2019-06-25 12:00
        </dd>
      </dl>
    </div>
  </article>
</section>
`;
